fluent/fluent.go

```diff
@@ -219,45 +241,92 @@ func e(x, y float64) int {
 	return int(math.Pow(x, y))
 }
 
-func (f *Fluent) reconnect() {
-	go func() {
-		for i := 0; ; i++ {
-			err := f.connect()
-			if err == nil {
-				f.mu.Lock()
-				f.reconnecting = false
-				f.mu.Unlock()
-				break
-			} else {
-				if i == f.Config.MaxRetry {
-					panic("fluent#reconnect: failed to reconnect!")
-				}
-				waitTime := f.Config.RetryWait * e(defaultReconnectWaitIncreRate, float64(i-1))
-				time.Sleep(time.Duration(waitTime) * time.Millisecond)
-			}
+func (f *Fluent) reconnect() error {
+	for i := 0; ; i++ {
+		err := f.connect()
+		if err == nil {
+			return nil
 		}
-	}()
+		if i == f.Config.MaxRetry {
+			return errors.New("fluent#reconnect: failed to reconnect!")
+		}
+		waitTime := f.Config.RetryWait * e(defaultReconnectWaitIncreRate, float64(i-1))
+		time.Sleep(time.Duration(waitTime) * time.Millisecond)
+	}
 }
 
 func (f *Fluent) flushBuffer() {
-	f.mu.Lock()
-	defer f.mu.Unlock()
-	f.pending = f.pending[0:0]
+	f.buf = f.buf[0:0]
 }
 
-func (f *Fluent) send() (err error) {
-	if f.conn == nil {
-		if f.reconnecting == false {
-			f.mu.Lock()
-			f.reconnecting = true
-			f.mu.Unlock()
-			f.reconnect()
+func (f *Fluent) send(data []byte) (err error) {
+	for {
+		if f.conn == nil {
+			if err = f.reconnect(); err != nil {
+				return
+			}
+		}
+		if _, err = f.conn.Write(data); err == nil {
```

Comment:
Is conn can be a nil pointer some times. make app panic. This problem happend in latest version.